paddle/framework/scope.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <unordered_map>
+#include <vector>
+#include "paddle/framework/variable.h"
+#include "paddle/utils/Error.h"
+
+namespace paddle {
+namespace framework {
+
+const static Error AlreadyCreated("Variable has already been created");
+
+/**
+ * Scope is an association of a name to Variable. All variables belong to
+ * `Scope`. You need to specify a scope to run a Net, i.e., `net.Run(&scope)`.
+ * One net can run in different scopes and update different variable in the
+ * scope.
+ */
+class Scope {
+ public:
+  Scope() {}
+
+  explicit Scope(const std::shared_ptr<Scope>& scope) : parent_(scope) {}
+
+  ~Scope() {}
+
+  // Create Variable in this Scope. Return error if Variable already been
+  // created.
+  Error __must_check CreateVariable(const std::string& name);
```

Comment:
Good argue!

Let us think about this following the first principle.

What users need are 

1. Create
1. Get

Create could err if there has been a variable with the same name there.  We can do three things with this error:

1. we do nothing
1. return the error

A basic design principle is that we handle the error as much as we could at where it happens, and try not to return to the users, because the users knows nothing about the detail and cannot really do much.  Also, if the variable is already there, users' intention (create the variable) is fulfilled, there is really nothing worth of returning.  So I prefer Create takes no-op with the duplication error.

Similarly, Get could err if it cannot find the variable, and we can take the following possible actions:

1. no-op
1. return error
1. create the variable

When the user Gets a variable, it assumes that the variable should have been there. If it has not, it implies that the creation process failed or we were creating a wrong network description. Anyway, there do have an error so we cannot take no-op.  Similarly, if we create the variable, we hide the error; this is not right. So we must report error. Fortunately, reporting the error is simple -- just return nullptr would be enough.

Summarize above derivation, I think we should have

1. CreateVar, which does nothing if there has been the variable, and
1. Get, which returns nullptr if there has not been the variable.